compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Validate.hs

```diff
@@ -0,0 +1,71 @@
+-- Copyright (c) 2020 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Daml.Compiler.Validate (validateDar) where
+
+import Control.Monad (forM_)
+import qualified "zip-archive" Codec.Archive.Zip as ZipArchive
+import qualified DA.Daml.LF.Ast as LF
+import qualified DA.Daml.LF.Proto3.Archive as Archive
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.NameMap as NM
+
+import DA.Daml.Compiler.ExtractDar (extractDar,ExtractedDar(..))
+import DA.Daml.LF.Ast.Version
+import DA.Daml.LF.Ast.World (initWorld)
+import DA.Pretty (renderPretty)
+import qualified DA.Daml.LF.TypeChecker as TC (checkModule)
+import qualified DA.Daml.LF.TypeChecker.Error as TC
+
+-- | Monad for performing Validation
+newtype Validation a = Validation (Either ValidationError a) deriving (Functor,Applicative,Monad)
+
+data ValidationError
+  = VeArchiveError FilePath Archive.ArchiveError
+  | VeTypeError TC.Error
+
+validationError :: ValidationError -> Validation a
+validationError = Validation . Left
+
+-- | Run a Validation, throwing an error when the validation fails
+runValidation :: Validation a -> a
+runValidation = \case
+  Validation (Left e) -> error (show e)
+  Validation (Right a) -> a
+
+instance Show ValidationError where
+  show = \case
+    VeArchiveError fp err -> unlines
+      [ "Invalid DAR."
+      , "DALF entry cannot be decoded: " <> fp
+      , show err ]
+    VeTypeError err -> unlines
+      [ "Invalid DAR."
+      , "The DAR is not well typed."
+      , renderPretty err ]
+
+-- | Validate a loaded DAR: that all DALFs are well-typed and consequently that the DAR is closed
+validateDar :: FilePath -> IO Int
+validateDar inFile = do
+  ExtractedDar{edDalfs} <- extractDar inFile
+  return $ runValidation $ do
```

Comment:
That might run into the exact laziness problem that I mentioned above (could be that you end up forcing it somehow atm but better be on the safe side and throw an IO exception via `errorIO`).